### ISSUE TITLE
Figure out how to return a custom error from `Guard`s

### ISSUE THREAD

**Issue by jonas-schievink** [dac-gmbh/hyperdrive#22]
Also needs an example/doctest.

**Comment by rustonaut** [comment]
I will figure out how to return a custom error first (not just from the guard but also from the route handler).

This will allow having custom HTTP error codes by having a service wrapper.

**Comment by rustonaut** [comment]
This would also mean changing the route handler to return a result which makes it more analogous to how hyper services generally work, I hope this is fine?

**Comment by jonas-schievink** [comment]
> This would also mean changing the route handler to return a result which makes it more analogous to how hyper services generally work, I hope this is fine?

Hmm, so technically you can already return whatever HTTP code you want in the `Response` you return, so this would be redundant. I originally didn't allow the handler to return a `Result` because it wouldn't be very clear what the semantics of returning an `Ok` with a response containing an HTTP error response would be. The intention was that the user would do their own `match` in the handler and turn any error into a `Response`.

A problem with the semantics of hyper's `Service` trait is that any error returned results in the connection being dropped immediately instead of an error being returned, and I don't want to do the same here. But even if we'd return an actual response here, the user can already do that in their own handler.

**Comment by rustonaut** [comment]
> A problem with the semantics of hyper's `Service` trait is that any error returned results in the connection being dropped immediately instead of an error being returned, and I don't want to do the same here. But even if we'd return an actual response here, the user can already do that in their own handler.

It turns out hyper does so for a good reason. It's basically used for errors of the protocol layer and lower caused by the client. E.g. physical disconnect or client starts sending garbage. It that happens the connection has "gone bad" so you can't really respond with an error code on the same time you don't want to log it as a server error because this would mean anyone could easily "DOS the log" (or you just get a lot of spam in the log if it's less malicious). Through you still might want to see it in metrics but that isn't supported for now.

The problem is that even in a normal HTTP services (even in a router function) you still can run into this kind of errors (e.g. when stream processing the input) so you want to allow the user to return this error if they need to.

Also if service errors where only meant for this it doesn't make sense that they are generic (you just would make them `BoxedError` or a wrapper around `BoxedError`). While hyper currently has no build in support for this kind of thinks by wrapping a service you can easily get it.

I happened to have a wrapper which does:

1. catch panics, and calls some `ErrorHandler.handle_panic()` function (which returns  a `Result<Response<Body>, BoxedError>`.
2. handles errors with a `Errorhandler.handle_error()` function (which also returns a `Result<Response<Body>, BoxedError>`).

So the changes I have in mind would _allow_ a sync route handler to return a `Result<Response<Body>, SomeGenericError>`. _But it also allows to still use functions which just return a `Response<Body>`_. Async route handlers already return a result, so it would just change the error from `BoxedError` to a generic error.

Combining this with a custom hyper service wrapper  allows sending error responses from `Guard`'s without entering the route handler.

# Summary of Pros/Cons

Benefits of it:
- Allow returning hyper errors for the view use-cases where it's needed and can't be put into the `FromBody` impl.
- Work with any service wrappers which add better error handling to hyper. ( The backend likes that ;=) )

Drawbacks:
- More complex in the implementation.

Thinks which don't change:
- Still allows returning just a `Response<Body>` for SyncService route handlers.
- `AsyncService` did return a error anyway, but now allows having custom errors, so not really changed
  (but maybe some cases where now a type hint is needed).

# PS

>  I originally didn't allow the handler to return a Result because it wouldn't be very clear what the semantics of returning an Ok with a response containing an HTTP error response would be.

The semantics would be:
- `Ok` => a valid response exists (which on a business/application level might be a error response but on hyper/service/protocol level isn't).
- `Err` => no valid response exists, but if some error handling is done (e.g. through wrapping the service) it might still be produced in another layer.

> The intention was that the user would do their own match in the handler and turn any error into a Response.

This isn't enough if you use any "wrapping" of services to e.g. properly handle panics with a 500 Internal Server Error.

# PPS

I will also open a Issue on hyper to  ask them what their intentions wrt. to panic and better error handling is.

**Comment by rustonaut** [comment]
Implementation wise I would have some marker trait with wild card impl. for all supported function signatures, which then is accepted by `SyncService` and `AsyncService` this would have some other beneficial side effects:

- Make it easier to support bodies outside of `Body`.
- Not only supports `fn`'s but also `Fn`'s 
- Also allows custom structs I think (which is interesting for very complex endpoints, but not really needed as far as I can tell, just a side effect that you now can do it if you want).

**Comment by jonas-schievink** [comment]
> It turns out hyper does so for a good reason. It's basically used for errors of the protocol layer and lower caused by the client. E.g. physical disconnect or client starts sending garbage.

Yeah, this does make sense, but at the point where Hyperdrive is involved the HTTP request is already fully parsed, except the body, so any parsing errors in hyper itself would've already terminated the connection (which I agree is reasonable to do there). If the body contains invalid data I'd expect the `FromBody` impl to return an error, which Hyperdrive currently turns into a 400 Bad Request (which I also think is the right thing to do by default; an application might want to change this or provide a response body though).


> Also if service errors where only meant for this it doesn't make sense that they are generic (you just would make them `BoxedError` or a wrapper around `BoxedError`). While hyper currently has no build in support for this kind of thinks by wrapping a service you can easily get it.

It does avoid having to convert to `BoxedError` inside every implementation though, so it's not completely useless. I think I'm on board with using generic errors instead of `BoxedError`s everywhere though (it might make Hyperdrive even more difficult to use though).



> Implementation wise I would have some marker trait with wild card impl. for all supported function signatures, which then is accepted by `SyncService` and `AsyncService`

Note that you can't have 2 blanket impls for a `T: Fn` with different parameters or return types (rustc always considers them to overlap – this is a known limitation in Rust right now). For example this doesn't work:

```rust
trait Tr {}

impl<T> Tr for T where T: Fn() -> u8 {}
impl<T> Tr for T where T: Fn(()) -> u16 {}
```

**Comment by rustonaut** [comment]
> > Implementation wise I would have some marker trait with wild card impl. for all supported function signatures, which then is accepted by SyncService and AsyncService
>
> Note that you can't have 2 blanket impls for a T: Fn with different parameters or return types (rustc always considers them to overlap – this is a known limitation in Rust right now). For example this doesn't work: [...]

Wait, but I made it compile with a overlap.....

Also there are still some cases where you body isn't fully handled by the `FromBody` impl and you can still run into hyper Errors in the route, mainly if you do some streaming of input data.
